stellar-contract-env-host/src/host.rs

```diff
@@ -1092,17 +1092,51 @@ impl CheckedEnv for Host {
 
     fn map_prev_key(&self, m: Object, k: RawVal) -> Result<RawVal, HostError> {
         let k = self.associate_raw_val(k);
-        self.visit_obj(m, |hm: &HostMap| match hm.get_prev(&k) {
-            Some((pk, pv)) => Ok(pk.to_raw()),
-            None => Ok(UNKNOWN_ERROR.to_raw()), //FIXME: replace with the actual status code
+        self.visit_obj(m, |hm: &HostMap| {
+            if let Some((pk, pv)) = hm.get_prev(&k) {
+                if *pk != k {
+                    Ok(pk.to_raw())
+                } else {
+                    if let Some((pk2, pv2)) = hm
+                        .clone()
+                        .extract(pk)
+                        .ok_or_else(|| HostError::General("key not exist"))?
+                        .1
+                        .get_prev(pk)
+                    {
+                        Ok(pk2.to_raw())
+                    } else {
+                        Ok(UNKNOWN_ERROR.to_raw()) //FIXME: replace with the actual status code
+                    }
+                }
+            } else {
+                Ok(UNKNOWN_ERROR.to_raw()) //FIXME: replace with the actual status code
+            }
         })
     }
 
     fn map_next_key(&self, m: Object, k: RawVal) -> Result<RawVal, HostError> {
         let k = self.associate_raw_val(k);
-        self.visit_obj(m, |hm: &HostMap| match hm.get_next(&k) {
-            Some((pk, pv)) => Ok(pk.to_raw()),
-            None => Ok(UNKNOWN_ERROR.to_raw()), //FIXME: replace with the actual status code
+        self.visit_obj(m, |hm: &HostMap| {
+            if let Some((pk, pv)) = hm.get_next(&k) {
+                if *pk != k {
+                    Ok(pk.to_raw())
+                } else {
+                    if let Some((pk2, pv2)) = hm
+                        .clone()
+                        .extract(pk)
+                        .ok_or_else(|| HostError::General("key not exist"))?
+                        .1
+                        .get_next(pk)
```

Comment:
How do these lines of code work? I'm struggling to see what is being extracted. This block of code might benefit from a comment I think.

Reply:
I added a comment. Let me know if it still needs more explanation.